analysis/filter/StructureModifier/src/main/kotlin/de/maibornwolff/codecharta/filter/structuremodifier/StructureModifier.kt

```diff
@@ -0,0 +1,105 @@
+        description = ["changes the structure of cc.json files"],
+        footer = ["Copyright(c) 2019, MaibornWolff GmbH"])
+class StructureModifier(private val input: InputStream = System.`in`,
+                        private val output: PrintStream = System.out,
+                        private val error: PrintStream = System.err) : Callable<Void?> {
+
+    @CommandLine.Option(names = ["-h", "--help"], usageHelp = true, description = ["displays this help and exits"])
+    var help: Boolean = false
+
+    @CommandLine.Parameters(arity = "0..1", paramLabel = "FILE", description = ["input project file"])
+    private var source: File? = null
+
+    @CommandLine.Option(names = ["-s", "--setRoot"], description = ["path within project to be extracted"])
+    private var setRoot: String? = null
+
+    @CommandLine.Option(arity = "1", names = ["-p", "--printLevels"], description = ["show first x layers of project hierarchy"])
+    private var printLevels: Int = 0
+
+    @CommandLine.Option(names = ["-o", "--outputFile"], description = ["output File (or empty for stdout)"])
+    private var outputFile: File? = null
+
+    @CommandLine.Option(names = ["-n", "--projectName"], description = ["project name of new file"])
+    private var projectName: String? = null
+
+    @CommandLine.Option(names = ["-f", "--moveFrom"], description = ["move nodes in project folder..."])
+    private var moveFrom: String? = null
+
+    @CommandLine.Option(arity = "1..*", names = ["-r", "--remove"], description = ["node(s) to be removed"])
+    private var remove: Array<String> = arrayOf()
+
+    @CommandLine.Option(names = ["-t", "--moveTo"], description = ["... move nodes to destination folder"])
+    private var moveTo: String? = null
+
+    private lateinit var project: Project
+    private val logger = KotlinLogging.logger {}
+
+    override fun call(): Void? {
+
+        project = readProject() ?: return null
+
+        when {
+            printLevels > 0 -> {
+                ProjectStructurePrinter(project, output).printProjectStructure(printLevels)
+                return null
+            }
+            setRoot != null -> project = SubProjectExtractor(project, projectName).extract(setRoot!!)
+            remove.isNotEmpty() -> project = NodeRemover(project, projectName).remove(remove)
+            moveFrom != null -> project = FolderMover(project, projectName).move(moveFrom, moveTo) ?: return null
+        }
+
+        val writer = outputFile?.bufferedWriter() ?: output.bufferedWriter()
+        ProjectSerializer.serializeProject(project, writer)
+
+        return null
+    }
+
+    private fun readProject(): Project? {
+        val bufferedReader = source?.bufferedReader() ?: return readPipedProject()
+        return try {
+            ProjectDeserializer.deserializeProject(bufferedReader)
+        } catch (e: Exception) {
+            val input = source!!.name
+            logger.error("$input is not a valid project file and is therefore skipped.")
```

Comment:
Where do you validate the project? Doesn't the `deserializeProject(...)` just check if its a valid json content? Also see line 87 in `readPipedProject()`

Reply:
good point, we have a Validation tool in ccsh (no idea how powerful it is), maybe it would be a good idea to test the input file with it. I will check that

Reply:
it validates against a json schema file `cc.json`

Reply:
i noticed, that the projectDeserializer actually checks for the structure of the input file, a further check is therefore not necessary. I added a testcase to verify that we are discarding jsons that are not valid project files.

Reply:
Ah nice, so this `GSON.fromJson(reader, Project::class.java)` actually compares the json input to the `Project`'s class structure 👍